### ISSUE TITLE
Memories scans same shared folder for several users

### ISSUE THREAD

**Issue by wonx** [pulsejet/memories#454]
This is somewhat related to issue https://github.com/pulsejet/memories/issues/184

*Memories* scans all files for all users. But if a particular folder is shared for several users, it will be scanned again for each user, despite already being scanned and the contents being the same. I have a media/document library of more than ~500000 files, which is shared (as a nextcloud shared folder) between three users, and memories is scanning it for the third time already.

For instance, I have my `files/user1/Media/Photos` folder, that belongs to `user1`. That Photos folder is shared with `user2` and `user3`, so they see it as `files/user2/Shared/Photos` and `files/user3/Shared/Photos` respectively. 

I think it's something that could be optimized.

**Comment by pulsejet** [comment]
tl;dr it doesn't matter.

If the same file is scanned for the second time, it is ignored and not re-indexed. All the files will be scanned only for the first user and only the timestamp will be compared over the rest. This should be very fast (a few minutes for 500000 files probably). Secondly, you only need to run `memories:index` once; subsequently uploaded files are automatically indexed.

As such, optimizing this would be premature.

**Comment by wonx** [comment]
Ok, maybe in my case the situation is a bit different. In addition to what I said in #`, *memories* is indexing redundant files because the files are in an External storage, and that external storage is  assigned to several users. Does that make sense?

So maybe restrict the amount of times *memories* goes through the same external storage, even if several users have access to it? In my case, scanning the whole libraries is taking several days (would be weeks if I didn't add a bunch of `.nomedia` files)

**Comment by pulsejet** [comment]
How many users do you have? Each file has a unique fileid. When a fileid has been indexed already, it is skipped directly (regardless of whether it is on external storage or not). It's downloaded only if it's never been indexed / newer. Are you sure the files are actually being re-indexed again (what's the output of the index command)?

The solution to this would be equivalent to #184. There's no way to identify if a storage has been scanned already; only files. So the solution would be to restrict scanning directories.

**Comment by wonx** [comment]
3 users, plus a test user where I assign all the unused external storages that I'm currently testing.

So right now `php occ memories:index` is going through the same files it did for user2, but for user3. I can tell it's indexing again because it's taking like half a second per file (If I stop the command and start it again, it goes much faster over the files which have alredy been indexed).

Another complementary solution would be the option to exclude certain users from memories. For instance, I have the `test` user which has lots of files assigned, and that user will never use memories, so it's taking hours to index for information that will never be used.

**Comment by pulsejet** [comment]
> I can tell it's indexing again because it's taking like half a second per file

This is either a bug, or these files exist as separate entires in the filecache (i.e. bug in Nextcloud or some config issue).

https://github.com/pulsejet/memories/blob/4bbf8770b7c6e33873891496399a444f4a0f5598/lib/Db/TimelineWrite.php#L106

You're not running index with `-f` right?

**Comment by wonx** [comment]
No, just `php occ memories:index`.

I don't have much idea on how data in Nextcloud is organized, but I checked the `oc_filecache` table in SQL and I found that files in the same external storage have a different `fileid` for each user.

For instance, this is a photo from an external storage assigned to two users.

`SELECT * FROM oc_filecache WHERE name = 'DSC00221_2.JPG';`

```
| fileid  | storage | path                                   | path_hash                        | parent  | name           | mimetype | mimepart | size  | mtime      | storage_mtime | encrypted | unencrypted_size | etag          | permissions | checksum |
|---------|---------|----------------------------------------|----------------------------------|---------|----------------|----------|----------|-------|------------|---------------|-----------|------------------|---------------|-------------|----------|
| 824037  | 9       | Fotos/2005-12 Amsterdam/DSC00221_2.JPG | 6df22fc3d76f6a1a0e86195dcd2f4839 | 799068  | DSC00221_2.JPG | 13       | 11       | 82032 | 1573594537 | 1573594537    | 0         | 0                | 5fd67990d25a9 | 27          |          |
| 4138530 | 30      | Fotos/2005-12 Amsterdam/DSC00221_2.JPG | 6df22fc3d76f6a1a0e86195dcd2f4839 | 4135359 | DSC00221_2.JPG | 13       | 11       | 82032 | 1573594537 | 1573594537    | 0         | 0                | 624cb445c572a | 27          |          |	

```

**Comment by pulsejet** [comment]
Yeah so in this case there's essentially no way for memories to know it's the same file, so it's gonna scan it again regardless.
